cirq-ft/cirq_ft/algos/phase_estimation.py

```diff
@@ -0,0 +1,81 @@
+# Copyright 2023 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import abc
+from cirq._compat import cached_property
+import cirq
+
+from cirq_ft import infra
+
+
+class PhaseEstimation(infra.GateWithRegisters):
+    r"""Abstract base class that defines the API for a Phase Estimation procedure"""
+
+    @property
+    @abc.abstractmethod
+    def control_register(self) -> infra.Register:
+        ...
+
+    @property
+    @abc.abstractmethod
+    def eigenvector_register(self) -> infra.Register:
+        ...
+
+    @property
+    @abc.abstractmethod
+    def control_register_prep(self) -> infra.GateWithRegisters:
+        ...
+
+    @property
+    @abc.abstractmethod
+    def eigen_vector_register_prep(self) -> infra.GateWithRegisters:
+        ...
+
+    @cached_property
+    def control_register_name(self):
+        return self.control_register.name
+
+    @cached_property
+    def eigenvector_register_name(self):
+        return self.eigenvector_register.name
+
+    @property
+    @abc.abstractmethod
+    def lamda(self) -> infra.GateWithRegisters:
```

Comment:
I used the name Lamda as per the University of Waterloo's convention for this gate.